### PR TITLE
feat(core): Sidecar Node API

### LINKED ISSUE #321: "Sidecar" nodes API.
## User Story

As a developer
I want to be able to add content nodes without changing the component hierarchy
So that I can release enhancements which do not require migration

## Acceptance Criteria

AC1. There is a core bodiless API which allows a HOC to use a content node hierarchy which starts with a child node of the wrapped component.
AC2. A component wrapped with an HOC utilizing the above API will receive the same content node as if it had not been wrapped.

### PR DESCRIPTION
**Pull request by wodenx**
## Changes

Adds an API for creating "sidecar" node hierarchies,
effectively allowing an enhancing HOC to hang its
data off of the component it wraps without changing
the node that is passed to that component.

## Test Instructions

See docs and functional tests.

## Related Issues
#285
Closes #321

### CODE REVIEW

File: packages/bodiless-core/src/withSidecarNodes.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
+
```
Comment (VancheeZze) [inline packages/bodiless-core/src/withSidecarNodes.tsx:1]: Missing copyright

File: packages/bodiless-core/__tests__/withSidecarNodes.test.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import React, { ComponentType } from 'react';
```
Comment (VancheeZze) [inline packages/bodiless-core/__tests__/withSidecarNodes.test.tsx:1]: Missing copyright text

File: packages/bodiless-documentation/doc/Development/Architecture/Data.md [hunk 3]
```diff
@@ -120,50 +123,180 @@ export const query = graphql`
...
+...
+<Blurb1 />
+```
+is equivalent to
+```
+<Blurb nodeKey="blurb-1" />
+```
+
+This pattern becomes particularly powerful when enhancing a compound component
+using he [Design API](../FClasses). Assuming we made blurb "designable", we
```
Comment (beliayeu) [inline packages/bodiless-documentation/doc/Development/Architecture/Data.md:289]: he -> the

File: packages/bodiless-documentation/doc/Development/Architecture/Data.md [hunk 4]
```diff
@@ -120,50 +123,180 @@ export const query = graphql`
...
+  blurb-1$title.json
+  blurb-1$body.json
+  blurb-2$title.json
+  blurb-2$body.json
+```
+
+### "Sidecar" nodes.
+
+There are some cases in which having the data model mirror the component hierarchy
+can be undesireable. For example, suppose we wanted to enhance our component above
```
Comment (beliayeu) [inline packages/bodiless-documentation/doc/Development/Architecture/Data.md:215]: undesireable -> undesirable

File: packages/bodiless-core/src/withSidecarNodes.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import React, { createContext, useContext, ComponentType } from 'react';
+import { flowRight } from 'lodash';
+import NodeProvider, { useNode } from './NodeProvider';
+import { ContentNode, DefaultContentNode } from './ContentNode';
+import withNode from './withNode';
+
+const SidecarNodeContext = createContext<ContentNode<any>>(DefaultContentNode.dummy());
+
+const startSidecarNodes = <P extends object>(Component: ComponentType<P>) => (props: P) => (
+  <SidecarNodeContext.Provider value={useNode().node}>
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/src/withSidecarNodes.tsx]: fixed.

File: packages/bodiless-core/src/withSidecarNodes.tsx [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    <Component {...props} />
+  </SidecarNodeContext.Provider>
+);
+
+const endSidecarNodes = <P extends object>(Component: ComponentType<P>) => (props: P) => (
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/src/withSidecarNodes.tsx]: fixed

File: packages/bodiless-core/__tests__/withSidecarNodes.test.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    );
+    const Test = asTestWithEnhancement(NodePathPrinter);
+    const wrapper = mount(<Test id="test" />);
+    const span = wrapper.find('span#test');
+    expect(span.text()).toBe('root$test');
+    expect(span.prop('data-enh-foo')).toBe('root$foo');
+    expect(span.prop('data-enh-bar')).toBe('root$foo$bar');
+  });
+});
+
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/__tests__/withSidecarNodes.test.tsx]: fixed

File: packages/bodiless-core/__tests__/withSidecarNodes.test.tsx [hunk 8]
```diff
@@ -17,121 +23,201 @@ const withEnhancement = (id: string) => (Component: ComponentType<any>) => withN
...
+        mockNodeData.mockReset();
+        mockNodeData.mockReturnValueOnce(true);
+        const wrapper = mount(<Test />);
+        expect(mockNodeData.mock.calls[0][0]).toBe('root$toggle');
+        expect(wrapper.find('span#test')).toHaveLength(0);
+        expect(wrapper.find('span#bar').prop('data-enh-bar')).toBe('root$foo');
+      });
+    });
+  });
+  describe('Enhancing a flow container item wrapper without changing data', () => {
```
Comment (beliayeu) [inline packages/bodiless-core/__tests__/withSidecarNodes.test.tsx:144]: are we fine with using flow container term inside bodiless/core?

File: packages/bodiless-core/__tests__/withSidecarNodes.test.tsx (hunk 2 above)
Comment (wodenx) [reply packages/bodiless-core/__tests__/withSidecarNodes.test.tsx:15]: fixed

File: packages/bodiless-core/src/withSidecarNodes.tsx (hunk 1 above)
Comment (wodenx) [reply packages/bodiless-core/src/withSidecarNodes.tsx:15]: fixed

File: packages/bodiless-documentation/doc/Development/Architecture/Data.md (hunk 3 above)
Comment (wodenx) [reply packages/bodiless-documentation/doc/Development/Architecture/Data.md:289]: fixed

File: packages/bodiless-documentation/doc/Development/Architecture/Data.md (hunk 4 above)
Comment (wodenx) [reply packages/bodiless-documentation/doc/Development/Architecture/Data.md:215]: fixed

File: packages/bodiless-core/__tests__/withSidecarNodes.test.tsx [hunk 9]
```diff
@@ -17,121 +23,201 @@ const withEnhancement = (id: string) => (Component: ComponentType<any>) => withN
...
+  withNode,
+  withPropEnhancement('baz'),
+);
 
-describe('withSidecarNodes', () => {
-  it('is tested against the correct baseline', () => {
-    const asTestWithoutEnhancement: HOC = flowRight(
-      withNodeKey('test'),
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const withWrapperEnhancement = (id: string) => (Component: ComponentType<any>) => (
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/__tests__/withSidecarNodes.test.tsx:62]: fixed

File: packages/bodiless-core/__tests__/withSidecarNodes.test.tsx (hunk 8 above)
Comment (wodenx) [reply packages/bodiless-core/__tests__/withSidecarNodes.test.tsx:144]: i think it's ok.

File: packages/bodiless-documentation/doc/Development/Architecture/Data.md [hunk 10]
```diff
@@ -120,50 +123,180 @@ export const query = graphql`
...
+provides a base ContentNode instance in to the Context. Components that need a
+node should wrap themselves in `withNode`. When supplied with a `nodeKey` prop,
+this will provide each of them with a new `ContentNode` identified by that key.
+The node can then be accessed via the `useNode` hook.
 
 ```javascript
+const ComponentWithData = withNode(
+  props => {
+    const { node } = useNode();
+    return (<span {,,,props}>{node.data}</span>);
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-documentation/doc/Development/Architecture/Data.md:142]: fixed

File: packages/bodiless-documentation/doc/Development/Architecture/Data.md [hunk 11]
```diff
@@ -255,26 +256,60 @@ placed on a site and was rendering actual data. When replaced with the enhanced
...
+
+Instead of defining our `ComponentWithData` as a Component, we can abstract that functionality
+to an HOC (a very common pattern in Bodiless-JS).
+```
+const asComponentWithData = BaseComponent => {
+  const ComponentWIthData = props => (
+    <BaseComponent {...props}>
+      {useNode().node.data}
+    </BaseComponent>
+};
```
Comment (beliayeu) [inline packages/bodiless-documentation/doc/Development/Architecture/Data.md:303]: return is missed

File: packages/bodiless-documentation/doc/Development/Architecture/Data.md [hunk 12]
```diff
@@ -2,7 +2,7 @@
 
 ## Overview
 
-Most components will consume and render some *data* (content and/or
+Most components will consume and render some GG*data* (content and/or
```
Comment (beliayeu) [inline packages/bodiless-documentation/doc/Development/Architecture/Data.md:5]: GG ?

File: packages/bodiless-core/__tests__/withSidecarNodes.test.tsx [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+      ),
+      withBaz('baz'),
+    )(spanWithId('test'));
+    const wrapper = mount(<Test />);
+    const span = wrapper.find('span#test');
+    expect(span.prop('data-enh-foo')).toBe('root$foobar$foo');
+    expect(span.prop('data-enh-bar')).toBe('root$foobar$bar');
+    expect(span.prop('data-enh-baz')).toBe('root$baz');
+  });
+});
```
Comment (e2thex) [inline packages/bodiless-core/__tests__/withSidecarNodes.test.tsx:217]: @wodenx do you want to add a test that has depth in the side car like a `withSidecarNodes(flowRight(withNodeKey('bar'), withFoo('foo))`

File: packages/bodiless-core/src/withSidecarNodes.tsx [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+ *
+ * @see `withSidecarNodes`
+ *
+ * @param Component Any component which uses the Bodiless ContentNode system.
+ */
+const endSidecarNodes = <P extends object>(Component: ComponentType<P>) => {
+  const EndSidecarNodes = (props: P) => {
+    const oldValue = useContext(SidecarNodeContext);
+    if (oldValue.length === 0) return <Component {...props} />;
+    const newNode = oldValue[oldValue.length - 1];
```
Comment (e2thex) [inline packages/bodiless-core/src/withSidecarNodes.tsx:58]: @wodenx I spent way to much time trying to find a more elegant way to do this ... I failed

File: packages/bodiless-documentation/doc/Development/Architecture/Data.md [hunk 15]
```diff
@@ -120,50 +123,213 @@ export const query = graphql`
...
+  withNode,
+  withSidecarNodes(
+    withNodeKey('aria-label'),
+    withAriaLabel,
+  ),
+  withNodeKey('component'),
+)(ComponentWithData);
+```
+giving
+```
```
Comment (e2thex) [inline packages/bodiless-documentation/doc/Development/Architecture/Data.md:287]: seeems like something is missing here

- [ ]  finish up this example

Comment (wodenx) [reply packages/bodiless-documentation/doc/Development/Architecture/Data.md:287]: fixed

Comment (wodenx) [reply packages/bodiless-documentation/doc/Development/Architecture/Data.md:287]: fixed

File: packages/bodiless-core/__tests__/withSidecarNodes.test.tsx (hunk 13 above)
Comment (wodenx) [reply packages/bodiless-core/__tests__/withSidecarNodes.test.tsx:217]: yes - but don't have time now.

File: packages/bodiless-documentation/doc/Development/Architecture/Data.md (hunk 11 above)
Comment (wodenx) [reply packages/bodiless-documentation/doc/Development/Architecture/Data.md:307]: fixed

File: packages/bodiless-documentation/doc/Development/Architecture/Data.md (hunk 12 above)
Comment (wodenx) [reply packages/bodiless-documentation/doc/Development/Architecture/Data.md]: fixed

File: packages/bodiless-core/src/withSidecarNodes.tsx (hunk 14 above)
Comment (wodenx) [reply packages/bodiless-core/src/withSidecarNodes.tsx:58]: ok let's resolve this - we can try some more later.

**Pull request merged**